File: apps/breakout.py
from machine import Pin
from neopixel import NeoPixel
from time import sleep_ms, ticks_ms, ticks_diff
import random
import _screens as screens

NAME = "Breakout"
NUM_LEDS = 64

np = None
JOY_LEFT = None
JOY_RIGHT = None

PADDLE_LEN = 3
PADDLE_COLOR = (0, 40, 50)
BALL_COLOR   = (50, 50, 50)
LEVEL_COL    = (0, 50, 30)
LIFE_COL     = (55, 0, 0)

FRAME_MS = 50
PADDLE_SPEED = 0.50
INIT_LIVES = 3
LAUNCH_HOLD_MS = 600
DEFLECTION = 0.18
MAX_VX = 0.55
MAX_VY = 0.55

LEVELS = (
    {
        "speed": 0.30,
        "layout": (
            "########",
            "........",
            "........",
            "........",
            "........",
            "........",
            "........",
            "........",
        ),
    },
    {
        "speed": 0.34,
        "layout": (
            "########",
            "########",
            "........",
            "........",
            "........",
            "........",
            "........",
            "........",
        ),
    },
    {
        "speed": 0.38,
        "layout": (
            "#.#.#.#.",
            ".#.#.#.#",
            "#.#.#.#.",
            "........",
            "........",
            "........",
            "........",
            "........",
        ),
    },
    {
        "speed": 0.42,
        "layout": (
            "########",
            "########",
            "########",
            "........",
            "........",
            "........",
            "........",
            "........",
        ),
    },
    {
        "speed": 0.48,
        "layout": (
            "...##...",
            "..####..",
            ".######.",
            "########",
            "........",
            "........",
            "........",
            "........",
        ),
    },
)


def clear():
    for i in range(NUM_LEDS):
        np[i] = (0, 0, 0)


def px(col, row, color):
    if 0 <= col <= 7 and 0 <= row <= 7:
        np[row * 8 + col] = color


def parse_level(layout):
    bricks = {}
    for i, line in enumerate(layout):
        row = 7 - i
        for col in range(min(8, len(line))):
            if line[col] == "#":
                bricks[(col, row)] = True
    return bricks


def brick_color(row):
    if row == 7: return (55, 0, 0)
    if row == 6: return (50, 25, 0)
    if row == 5: return (45, 40, 0)
    if row == 4: return (0, 45, 0)
    return (0, 25, 50)


def draw_paddle(paddle_center):
    pc = int(round(paddle_center))
    for off in (-1, 0, 1):
        c = pc + off
        if 0 <= c <= 7:
            px(c, 0, PADDLE_COLOR)


def draw_bricks(bricks):
    for (c, r) in bricks:
        px(c, r, brick_color(r))


def render(paddle_center, ball_x, ball_y, bricks, ball_visible=True):
    clear()
    draw_bricks(bricks)
    draw_paddle(paddle_center)
    if ball_visible:
        bx = int(round(ball_x))
        by = int(round(ball_y))
        px(bx, by, BALL_COLOR)
    np.write()


def cap_speed(vx, vy):
    if vx > MAX_VX: vx = MAX_VX
    elif vx < -MAX_VX: vx = -MAX_VX
    if vy > MAX_VY: vy = MAX_VY
    elif vy < -MAX_VY: vy = -MAX_VY
    return vx, vy


def update_paddle(paddle_center):
    if JOY_LEFT.value() == 0:
        paddle_center -= PADDLE_SPEED
    if JOY_RIGHT.value() == 0:
        paddle_center += PADDLE_SPEED
    if paddle_center < 1.0:
        paddle_center = 1.0
    if paddle_center > 6.0:
        paddle_center = 6.0
    return paddle_center


def play_ball(bricks, base_speed, paddle_center, score):
    """Returns ('cleared'|'lost'|'exit', score, paddle_center)."""
    ball_x = paddle_center
    ball_y = 1.0
    ball_vx = 0.0
    ball_vy = 0.0

    t0 = ticks_ms()
    while ticks_diff(ticks_ms(), t0) < LAUNCH_HOLD_MS:
        if screens.check_exit():
            return ("exit", score, paddle_center)
        paddle_center = update_paddle(paddle_center)
        ball_x = paddle_center
        render(paddle_center, ball_x, ball_y, bricks)
        sleep_ms(FRAME_MS)

    ball_vy = base_speed
    ball_vx = (random.random() - 0.5) * 0.3

    while True:
        if screens.check_exit():
            return ("exit", score, paddle_center)

        paddle_center = update_paddle(paddle_center)

        ball_x += ball_vx
        ball_y += ball_vy

        if ball_x < 0:
            ball_x = -ball_x
            ball_vx = -ball_vx
        elif ball_x > 7:
            ball_x = 14 - ball_x
            ball_vx = -ball_vx

        if ball_y > 7:
            ball_y = 14 - ball_y
            ball_vy = -ball_vy

        bc = int(round(ball_x))
        br = int(round(ball_y))
        if 0 <= bc <= 7 and 0 <= br <= 7 and (bc, br) in bricks:
            del bricks[(bc, br)]
            score += 1
            prev_x = ball_x - ball_vx
            prev_y = ball_y - ball_vy
            crossed_y = int(round(prev_y)) != br
            crossed_x = int(round(prev_x)) != bc
            if crossed_y and not crossed_x:
                ball_vy = -ball_vy
            elif crossed_x and not crossed_y:
                ball_vx = -ball_vx
            else:
                ball_vy = -ball_vy
                ball_vx = -ball_vx

        if ball_y <= 0 and ball_vy < 0:
            if abs(ball_x - paddle_center) <= 1.5:
                ball_y = -ball_y
                ball_vy = -ball_vy
                offset = ball_x - paddle_center
                ball_vx += offset * DEFLECTION
                ball_vx, ball_vy = cap_speed(ball_vx, ball_vy)
            else:
                render(paddle_center, ball_x, ball_y, bricks, ball_visible=False)
                return ("lost", score, paddle_center)

        if not bricks:
            render(paddle_center, ball_x, ball_y, bricks)
            sleep_ms(200)
            return ("cleared", score, paddle_center)

        render(paddle_center, ball_x, ball_y, bricks)
        sleep_ms(FRAME_MS)


def level_clear_flash():
    for _ in range(2):
        for i in range(NUM_LEDS):
            np[i] = (0, 50, 30)
        np.write()
        sleep_ms(80)
        clear()
        np.write()
        sleep_ms(60)


def play_one_game():
    """Returns final score, or None if exit triggered mid-play."""
    score = 0
    lives = INIT_LIVES
    level_idx = 0
    paddle_center = 4.0

    while True:
        if screens.check_exit():
            return None

        if screens.show_digit_briefly(level_idx + 1, LEVEL_COL, 600) == "exit":
            return None

        bricks = parse_level(LEVELS[level_idx]["layout"])
        speed = LEVELS[level_idx]["speed"]

        while bricks and lives > 0:
            if screens.check_exit():
                return None

            if screens.show_digit_briefly(lives, LIFE_COL, 500) == "exit":
                return None

            outcome, score, paddle_center = play_ball(bricks, speed, paddle_center, score)
            if outcome == "exit":
                return None
            if outcome == "lost":
                lives -= 1
            elif outcome == "cleared":
                break

        if lives <= 0:
            return score

        level_clear_flash()
        level_idx = (level_idx + 1) % len(LEVELS)


def run(neopixel, joystick):
    global np, JOY_LEFT, JOY_RIGHT
    np = neopixel
    JOY_LEFT = joystick["left"]
    JOY_RIGHT = joystick["right"]
    screens.init(neopixel, joystick)
    while True:
        if screens.loading_screen() == "exit":
            return
        score = play_one_game()
        if score is None:
            return
        if screens.game_over_screen(score) == "exit":
            return


if __name__ == "__main__":
    _np = NeoPixel(Pin(19, Pin.OUT), NUM_LEDS)
    _joy = {
        "up":     Pin(3, Pin.IN),
        "down":   Pin(6, Pin.IN),
        "left":   Pin(7, Pin.IN),
        "right":  Pin(2, Pin.IN),
        "center": Pin(8, Pin.IN),
        "slide":  Pin(9, Pin.IN),
    }
    run(_np, _joy)

File: web-toolkit/src/lib/simulator/apps/breakout.ts
import type { DisplayDims, Joystick, NeoPixel, Pin, RGB } from "../types";
import * as screens from "../screens";
import { sleep_ms, ticks_diff, ticks_ms } from "../runtime/time";

export const NAME = "Breakout";

const NUM_LEDS = 64;

const PADDLE_COLOR: RGB = [0, 40, 50];
const BALL_COLOR: RGB = [50, 50, 50];
const LEVEL_COL: RGB = [0, 50, 30];
const LIFE_COL: RGB = [55, 0, 0];

const FRAME_MS = 50;
const PADDLE_SPEED = 0.5;
const INIT_LIVES = 3;
const LAUNCH_HOLD_MS = 600;
const DEFLECTION = 0.18;
const MAX_VX = 0.55;
const MAX_VY = 0.55;

interface Level { speed: number; layout: readonly string[]; }

const LEVELS: readonly Level[] = [
  { speed: 0.30, layout: [
    "########",
    "........", "........", "........",
    "........", "........", "........", "........",
  ]},
  { speed: 0.34, layout: [
    "########", "########",
    "........", "........", "........",
    "........", "........", "........",
  ]},
  { speed: 0.38, layout: [
    "#.#.#.#.", ".#.#.#.#", "#.#.#.#.",
    "........", "........", "........",
    "........", "........",
  ]},
  { speed: 0.42, layout: [
    "########", "########", "########",
    "........", "........", "........",
    "........", "........",
  ]},
  { speed: 0.48, layout: [
    "...##...", "..####..", ".######.", "########",
    "........", "........", "........", "........",
  ]},
];

let np: NeoPixel;
let JOY_LEFT: Pin, JOY_RIGHT: Pin;

function clear(): void {
  for (let i = 0; i < NUM_LEDS; i++) np[i] = [0, 0, 0];
}

function px(col: number, row: number, color: RGB): void {
  if (col >= 0 && col <= 7 && row >= 0 && row <= 7) {
    np[row * 8 + col] = color;
  }
}

type Bricks = Map<string, [number, number]>;

function parseLevel(layout: readonly string[]): Bricks {
  const bricks: Bricks = new Map();
  for (let i = 0; i < layout.length; i++) {
    const row = 7 - i;
    const line = layout[i];
    for (let col = 0; col < Math.min(8, line.length); col++) {
      if (line[col] === "#") bricks.set(`${col},${row}`, [col, row]);
    }
  }
  return bricks;
}

function brickColor(row: number): RGB {
  if (row === 7) return [55, 0, 0];
  if (row === 6) return [50, 25, 0];
  if (row === 5) return [45, 40, 0];
  if (row === 4) return [0, 45, 0];
  return [0, 25, 50];
}

function drawPaddle(paddleCenter: number): void {
  const pc = Math.round(paddleCenter);
  for (const off of [-1, 0, 1]) {
    const c = pc + off;
    if (c >= 0 && c <= 7) px(c, 0, PADDLE_COLOR);
  }
}

function drawBricks(bricks: Bricks): void {
  for (const [c, r] of bricks.values()) px(c, r, brickColor(r));
}

function render(paddleCenter: number, ballX: number, ballY: number, bricks: Bricks, ballVisible = true): void {
  clear();
  drawBricks(bricks);
  drawPaddle(paddleCenter);
  if (ballVisible) {
    const bx = Math.round(ballX);
    const by = Math.round(ballY);
    px(bx, by, BALL_COLOR);
  }
  np.write();
}

function capSpeed(vx: number, vy: number): [number, number] {
  if (vx > MAX_VX) vx = MAX_VX;
  else if (vx < -MAX_VX) vx = -MAX_VX;
  if (vy > MAX_VY) vy = MAX_VY;
  else if (vy < -MAX_VY) vy = -MAX_VY;
  return [vx, vy];
}

function updatePaddle(paddleCenter: number): number {
  if (JOY_LEFT.value() === 0) paddleCenter -= PADDLE_SPEED;
  if (JOY_RIGHT.value() === 0) paddleCenter += PADDLE_SPEED;
  if (paddleCenter < 1.0) paddleCenter = 1.0;
  if (paddleCenter > 6.0) paddleCenter = 6.0;
  return paddleCenter;
}

type BallOutcome = "cleared" | "lost" | "exit";

async function playBall(
  bricks: Bricks,
  baseSpeed: number,
  paddleCenter: number,
  score: number,
): Promise<{ outcome: BallOutcome; score: number; paddleCenter: number }> {
  let ballX = paddleCenter;
  let ballY = 1.0;
  let ballVx = 0.0;
  let ballVy = 0.0;

  const t0 = ticks_ms();
  while (ticks_diff(ticks_ms(), t0) < LAUNCH_HOLD_MS) {
    if (screens.check_exit()) return { outcome: "exit", score, paddleCenter };
    paddleCenter = updatePaddle(paddleCenter);
    ballX = paddleCenter;
    render(paddleCenter, ballX, ballY, bricks);
    await sleep_ms(FRAME_MS);
  }

  ballVy = baseSpeed;
  ballVx = (Math.random() - 0.5) * 0.3;

  while (true) {
    if (screens.check_exit()) return { outcome: "exit", score, paddleCenter };

    paddleCenter = updatePaddle(paddleCenter);

    ballX += ballVx;
    ballY += ballVy;

    if (ballX < 0) {
      ballX = -ballX;
      ballVx = -ballVx;
    } else if (ballX > 7) {
      ballX = 14 - ballX;
      ballVx = -ballVx;
    }

    if (ballY > 7) {
      ballY = 14 - ballY;
      ballVy = -ballVy;
    }

    const bc = Math.round(ballX);
    const br = Math.round(ballY);
    const key = `${bc},${br}`;
    if (bc >= 0 && bc <= 7 && br >= 0 && br <= 7 && bricks.has(key)) {
      bricks.delete(key);
      score += 1;
      const prevX = ballX - ballVx;
      const prevY = ballY - ballVy;
      const crossedY = Math.round(prevY) !== br;
      const crossedX = Math.round(prevX) !== bc;
      if (crossedY && !crossedX) ballVy = -ballVy;
      else if (crossedX && !crossedY) ballVx = -ballVx;
      else {
        ballVy = -ballVy;
        ballVx = -ballVx;
      }
    }

    if (ballY <= 0 && ballVy < 0) {
      if (Math.abs(ballX - paddleCenter) <= 1.5) {
        ballY = -ballY;
        ballVy = -ballVy;
        const offset = ballX - paddleCenter;
        ballVx += offset * DEFLECTION;
        [ballVx, ballVy] = capSpeed(ballVx, ballVy);
      } else {
        render(paddleCenter, ballX, ballY, bricks, false);
        return { outcome: "lost", score, paddleCenter };
      }
    }

    if (bricks.size === 0) {
      render(paddleCenter, ballX, ballY, bricks);
      await sleep_ms(200);
      return { outcome: "cleared", score, paddleCenter };
    }

    render(paddleCenter, ballX, ballY, bricks);
    await sleep_ms(FRAME_MS);
  }
}

async function levelClearFlash(): Promise<void> {
  for (let i = 0; i < 2; i++) {
    for (let j = 0; j < NUM_LEDS; j++) np[j] = [0, 50, 30];
    np.write();
    await sleep_ms(80);
    clear();
    np.write();
    await sleep_ms(60);
  }
}

async function playOneGame(): Promise<number | null> {
  let score = 0;
  let lives = INIT_LIVES;
  let levelIdx = 0;
  let paddleCenter = 4.0;

  while (true) {
    if (screens.check_exit()) return null;

    if ((await screens.show_digit_briefly(levelIdx + 1, LEVEL_COL, 600)) === "exit") return null;

    const bricks = parseLevel(LEVELS[levelIdx].layout);
    const speed = LEVELS[levelIdx].speed;

    while (bricks.size > 0 && lives > 0) {
      if (screens.check_exit()) return null;

      if ((await screens.show_digit_briefly(lives, LIFE_COL, 500)) === "exit") return null;

      const res = await playBall(bricks, speed, paddleCenter, score);
      score = res.score;
      paddleCenter = res.paddleCenter;
      if (res.outcome === "exit") return null;
      if (res.outcome === "lost") lives -= 1;
      else if (res.outcome === "cleared") break;
    }

    if (lives <= 0) return score;

    await levelClearFlash();
    levelIdx = (levelIdx + 1) % LEVELS.length;
  }
}

export async function run(
  neopixel: NeoPixel,
  joystick: Joystick,
  display?: DisplayDims,
  screensNp?: NeoPixel,
): Promise<void> {
  np = neopixel;
  JOY_LEFT = joystick.left;
  JOY_RIGHT = joystick.right;
  screens.init(screensNp ?? neopixel, joystick, display?.width, display?.height);
  while (true) {
    if ((await screens.loading_screen()) === "exit") return;
    const score = await playOneGame();
    if (score === null) return;
    if ((await screens.game_over_screen(score)) === "exit") return;
  }
}
